src/freenet/support/math/TrivialRunningAverage.java

```diff
@@ -1,77 +1,69 @@
 package freenet.support.math;
 
+import java.util.concurrent.atomic.AtomicReference;
+
 public final class TrivialRunningAverage implements RunningAverage, Cloneable {
 
-	private static final long serialVersionUID = 1L;
-	private long reports;
-	private double total;
-
-	/**
-	 *
-	 * @param average
-	 */
-	public TrivialRunningAverage(TrivialRunningAverage average) {
-		this.reports = average.reports;
-		this.total = average.total;
-	}
-
-	/**
-	 *
-	 */
-	public TrivialRunningAverage() {
-		reports = 0;
-		total = 0.0;
-	}
-
-	@Override
-	public synchronized long countReports() {
-		return reports;
-	}
-
-	public synchronized double totalValue() {
-		return total;
-	}
-
-        /**
-         *
-         * @return
-         */
-        @Override
-        public synchronized double currentValue() {
-		return total / reports;
-	}
-
-        /**
-         *
-         * @param d
-         */
-        @Override
-        public synchronized void report(double d) {
-		total += d;
-		reports++;
-		// TODO Auto-generated method stub
-	}
-
-        /**
-         *
-         * @param d
-         */
-        @Override
-        public void report(long d) {
-		report((double)d);
-	}
-
-	@Override
-	public synchronized double valueIfReported(double r) {
-		return (total + r) / (reports + 1);
-	}
-
-	@Override
-	public TrivialRunningAverage clone() {
-		// Override clone() for synchronization.
-		// Implement Cloneable to shut up findbugs.
-		synchronized (this) {
-			return new TrivialRunningAverage(this);
-		}
-	}
+    private final AtomicReference<Data> data = new AtomicReference<>();
+
+    public TrivialRunningAverage(TrivialRunningAverage other) {
+        data.set(other.data.get());
+    }
+
+    public TrivialRunningAverage() {
+        data.set(new Data(0, 0));
+    }
+
+    @Override
+    public long countReports() {
+        return data.get().reports;
+    }
+
+    public double totalValue() {
+        return data.get().total;
+    }
+
+    @Override
+    public double currentValue() {
+        return data.get().getRunningAverage();
+    }
+
+    @Override
+    public void report(double d) {
+        data.updateAndGet(data -> data.updated(d));
+    }
+
+    @Override
+    public void report(long d) {
+        report((double) d);
+    }
+
+    @Override
+    public double valueIfReported(double r) {
+        return data.get().updated(r).getRunningAverage();
+    }
+
+    @Override
+    public TrivialRunningAverage clone() {
+        return new TrivialRunningAverage(this);
+    }
+
+    private static class Data {
+        private final long reports;
+        private final double total;
+
+        Data(long reports, double total) {
+            this.reports = reports;
+            this.total = total;
+        }
+
+        Data updated(double d) {
+            return new Data(reports + 1, d + total);
```

Comment:
This seems very expensive. Do we really need it?

Reply:
This adds two numbers and constructs a tiny object. What exactly seems very expensive here? :)

> Do we really need the trivial running average?

It is used all over the place (46 locations) so we cannot just remove it.